File: February2022/Python/MergekSortedLists.py
# TC : O(N*logN) , Where N --> n(number of nodes)*k(lenght of lists)

# SC : O(N)

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        valList = []
        for list in lists:
            while list != None:
                valList.append(list.val)
                list = list.next
        valList.sort()
        head = dummy = ListNode(0)
        for nodeVal in valList:
            dummy.next = ListNode(nodeVal)
            dummy = dummy.next
        return head.next

File: February2022/Java/MergeKSortedLists.java
// Author : Shobhit Behl(shobhitbruh)
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> pq=new PriorityQueue<>();
        ListNode dummy=new ListNode(0);
        ListNode result=dummy;
        
        for(ListNode x:lists){
            ListNode temp=x;
            while(temp!=null){
                pq.add(temp.val);
                temp=temp.next;
            }
        }
        
        while(pq.size()>0){
            ListNode n=new ListNode(pq.remove());
            dummy.next=n;
            dummy=dummy.next;
        }
        
        return result.next;
    }
}